pkg/nodeagent/pod_controller.go

```diff
@@ -237,101 +242,122 @@ func buildRmdWorkload(pod *corev1.Pod) ([]*intelv1alpha1.RmdWorkload, error) {
+		containerInfo, err := getContainerInfo(pod, container)
 		if err != nil {
-			logger.Error(err, "failed to retrieve cpuset from cgroups")
 			return nil, err
 		}
-		if len(coreIDs) == 0 {
-			logger.Info("cpuset not found in cgroups for container")
-			return nil, nil
-		}
 
 		rmdWorkload := &intelv1alpha1.RmdWorkload{}
-		// Create workload name. Naming convention: "<pod-name>-rmd-workload-<container-name>"
+		//Create workload name. Convention: "<pod-name>rmd-workload-<container-name>"
 		podName := string(pod.GetObjectMeta().GetName())
 		rmdWorkloadName := fmt.Sprintf("%s%s%s", podName, rmdWorkloadNameConst, container.Name)
 		podNamespace := pod.GetObjectMeta().GetNamespace()
 		if podNamespace == "" {
 			podNamespace = defaultNamespace
 		}
-
 		rmdWorkloadNamespacedName := types.NamespacedName{
 			Name:      rmdWorkloadName,
 			Namespace: podNamespace,
 		}
 
 		rmdWorkload.SetName(rmdWorkloadNamespacedName.Name)
 		rmdWorkload.SetNamespace(rmdWorkloadNamespacedName.Namespace)
-
-		maxCache, err := getMaxCache(&container)
-		if err != nil {
-			return nil, err
-		}
-		rmdWorkload.Spec.Rdt.Cache.Max = maxCache
-		rmdWorkload.Spec.CoreIds = coreIDs
+		rmdWorkload.Spec.Rdt.Cache.Max = containerInfo.maxCache
+		rmdWorkload.Spec.CoreIds = containerInfo.coreIDs
 		rmdWorkload.Spec.Nodes = make([]string, 0)
 		rmdWorkload.Spec.Nodes = append(rmdWorkload.Spec.Nodes, pod.Spec.NodeName)
 
-		workloadData := pod.GetObjectMeta().GetAnnotations()
-		for field, data := range workloadData {
-			if !strings.HasPrefix(field, container.Name) {
-				continue
-			}
+		getAnnotationInfo(rmdWorkload, pod, container.Name) //Changes workload in getAnnotationInfo()
 
-			switch {
-			case strings.HasSuffix(field, policyConst):
-				if data != "" {
-					rmdWorkload.Spec.Policy = data
-				}
-			case strings.HasSuffix(field, cacheMinConst):
-				minCache, err := strconv.Atoi(data)
-				if err != nil {
-					continue
-				}
-				rmdWorkload.Spec.Rdt.Cache.Min = minCache
-			case strings.HasSuffix(field, mbaPercentageConst):
-				mbaPercentage, err := strconv.Atoi(data)
-				if err != nil {
-					continue
-				}
-				rmdWorkload.Spec.Rdt.Mba.Percentage = mbaPercentage
-			case strings.HasSuffix(field, mbaMbpsConst):
-				mbaMbps, err := strconv.Atoi(data)
-				if err != nil {
-					continue
-				}
-				rmdWorkload.Spec.Rdt.Mba.Mbps = mbaMbps
-			case strings.HasSuffix(field, pstateRatioConst):
-				if data != "" {
-					rmdWorkload.Spec.Plugins.Pstate.Ratio = data
-				}
-			case strings.HasSuffix(field, pstateMonitoringConst):
-				if data != "" {
-					rmdWorkload.Spec.Plugins.Pstate.Monitoring = data
-				}
-			}
-
-		}
 		rmdWorkloads = append(rmdWorkloads, rmdWorkload)
 	}
 	return rmdWorkloads, nil
 }
 
+func getAnnotationInfo(rmdWorkload *intelv1alpha1.RmdWorkload, pod *corev1.Pod, containerName string) {
+	var minCacheStr, mbaPercentStr, mbaMbpsStr string = "0", "0", "0" //Ensures strconv.Atoi() doesn't receive an empty string
```

Comment:
We should not need these. Instead we should be able to handle the error correctly. See next comment.